internal/xds/rbac/rbac_engine.go

```diff
@@ -45,38 +122,112 @@ func NewEngine(policy *v3rbacpb.RBAC) (*Engine, error) {
+	conn := getConnection(data.Ctx)
+	_, dPort, err := net.SplitHostPort(conn.LocalAddr().String())
+	if err != nil {
+		return nil, err
+	}
+	dp, err := strconv.ParseUint(dPort, 10, 32)
+	if err != nil {
+		return nil, err
+	}
+
+	var peerCertificates []*x509.Certificate
+	if pi.AuthInfo != nil {
+		tlsInfo, ok := pi.AuthInfo.(credentials.TLSInfo)
+		if ok {
+			peerCertificates = tlsInfo.State.PeerCertificates
+		}
+	}
+
+	return &rpcData{
+		md:              md,
+		peerInfo:        pi,
+		fullMethod:      data.MethodName,
+		destinationPort: uint32(dp),
+		destinationAddr: conn.LocalAddr(),
+		certs:           peerCertificates,
+	}, nil
+}
+
+// rpcData wraps data pulled from an incoming RPC that the RBAC engine needs to
 // find a matching policy.
-type RPCData struct {
-	// MD is the HTTP Headers that are present in the incoming RPC.
-	MD metadata.MD
-	// PeerInfo is information about the downstream peer.
-	PeerInfo *peer.Peer
-	// FullMethod is the method name being called on the upstream service.
-	FullMethod string
-	// DestinationPort is the port that the RPC is being sent to on the
+type rpcData struct {
+	// md is the HTTP Headers that are present in the incoming RPC.
+	md metadata.MD
+	// peerInfo is information about the downstream peer.
+	peerInfo *peer.Peer
+	// fullMethod is the method name being called on the upstream service.
+	fullMethod string
+	// destinationPort is the port that the RPC is being sent to on the
 	// server.
-	DestinationPort uint32
-	// DestinationAddr is the address that the RPC is being sent to.
-	DestinationAddr net.Addr
-	// PrincipalName is the name of the downstream principal. If set, the URI
-	// SAN or DNS SAN in that order is used from the certificate, otherwise the
-	// subject field is used. If unset, it applies to any user that is
-	// authenticated.
-	PrincipalName string
+	destinationPort uint32
+	// destinationAddr is the address that the RPC is being sent to.
+	destinationAddr net.Addr
+	// certs will be used for authenticated matcher.
+	certs []*x509.Certificate
 }
 
-// FindMatchingPolicy determines if an incoming RPC matches a policy. On a
-// successful match, it returns the name of the matching policy and a true
-// boolean to specify that there was a matching policy found.
-func (r *Engine) FindMatchingPolicy(data *RPCData) (string, bool) {
+var errPolicyNotFound = errors.New("a matching policy was not found")
+
+// findMatchingPolicy determines if an incoming RPC matches a policy. On a
+// successful match, it returns the name of the matching policy and a nil error
+// to specify that there was a matching policy found.  It returns an error in
+// the case of not finding a matching policy.
+func (r *engine) findMatchingPolicy(rpcData *rpcData) (string, error) {
 	for policy, matcher := range r.policies {
-		if matcher.match(data) {
-			return policy, true
+		if matcher.match(rpcData) {
+			return policy, nil
 		}
 	}
-	return "", false
+	return "", errPolicyNotFound
 }
```

Comment:
A comment about code organization. Currently this is what we have:

- `ChainEngine` definition, its constructor and methods
- `engine` defintion and its constructor
- `connectionKey` and functions to get and set it
- `Data`, `newRPCData` and `rpcData` definitions
- `errPolicyNotFound` defintion
- `engine` method

My proposal would be as follows:
- `ChainEngine` definition, its constructor and methods
- `engine` definition, its constructor and methods
- `rpcData` and its constructor
- `connectionKey` and functions to get and set it

Reply:
Good call. Switched to your suggestion.